tests/Unit/Product/ProductRepositoryTest.php

```diff
@@ -0,0 +1,272 @@
+ * @property ProductMetaHandler $product_meta
+ * @property ProductHelper      $product_helper
+ */
+class ProductRepositoryTest extends ContainerAwareUnitTest {
+
+	use ProductTrait;
+
+	public function test_find_returns_all_supported_products() {
+		// supported
+		$simple_product   = WC_Helper_Product::create_simple_product();
+		$variable_product = WC_Helper_Product::create_variation_product();
+
+		// unsupported
+		WC_Helper_Product::create_external_product();
+
+		$results = $this->product_repository->find();
+
+		$variations = array_map( 'wc_get_product', $variable_product->get_children() );
+		$this->assertCount( count( $variations ) + 2, $results );
+		$this->assertEqualSets(
+			array_merge(
+				[
+					wc_get_product( $simple_product->get_id() ),
+					wc_get_product( $variable_product->get_id() ),
+				],
+				$variations
+			),
+			$results
+		);
+	}
+
+	public function test_find_always_returns_objects() {
+		WC_Helper_Product::create_simple_product();
+		$results = $this->product_repository->find( [ 'return' => 'ids' ] );
+		$this->assertContainsOnlyInstancesOf( WC_Product::class, $results );
+	}
+
+	public function test_find_returns_limited_and_based_on_offset() {
+		WC_Helper_Product::create_simple_product();
+		WC_Helper_Product::create_variation_product();
+
+		$results_unlimited = $this->product_repository->find();
+
+		$results = $this->product_repository->find( [], 1, 1 );
+		$this->assertCount( 1, $results );
+		$this->assertEquals( [ wc_get_product( $results_unlimited[1]->get_id() ) ], $results );
+	}
+
+	public function test_find_ids_always_returns_integer_ids() {
+		WC_Helper_Product::create_simple_product();
+		$results = $this->product_repository->find_ids( [ 'return' => 'objects' ] );
+		$this->assertContainsOnly( 'integer', $results );
+	}
+
+	public function test_find_passes_query_args() {
+		$product_1 = WC_Helper_Product::create_simple_product();
+		$product_1->set_status( 'draft' );
+		$product_1->save();
+
+		$product_2 = WC_Helper_Product::create_simple_product();
+		$product_2->set_status( 'draft' );
+		$product_2->save();
+		$this->product_meta->update_errors( $product_2, [ 'Error' ] );
+
+		$product_3 = WC_Helper_Product::create_simple_product();
+		$product_3->set_status( 'publish' );
+		$product_3->save();
+
+		WC_Helper_Product::create_variation_product();
+
+		$query_args = [
+			'status'     => 'draft',
+			'type'       => [ 'simple' ],
+			'meta_query' => [
+				'relation' => 'OR',
+				[
+					'key'     => ProductMetaHandler::KEY_ERRORS,
+					'compare' => 'NOT EXISTS',
+				],
+				[
+					'key'     => ProductMetaHandler::KEY_ERRORS,
+					'compare' => '=',
+					'value'   => '',
+				],
+			],
+		];
+		$this->assertEquals( [ $product_1->get_id() ], $this->product_repository->find_ids( $query_args ) );
+		$this->assertEquals( [ wc_get_product( $product_1 ) ], $this->product_repository->find( $query_args ) );
+	}
+
+	public function test_find_by_ids() {
```

Comment:
In this test we only create two products, so `find` would be equal to `find_by_ids`. Do we want to add a few more products to make sure we only get back the ones we requested?

Reply:
Good call. Added in https://github.com/woocommerce/google-listings-and-ads/pull/855/commits/1dc146f57235c45fafe6dc1af524f42f292f0c09